massa-factory-worker/src/denunciation_factory.rs

```diff
@@ -0,0 +1,317 @@
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::thread;
+
+use crossbeam_channel::{select, Receiver};
+use tracing::{debug, info, warn};
+
+use massa_factory_exports::{FactoryChannels, FactoryConfig};
+use massa_models::address::Address;
+use massa_models::denunciation::{Denunciation, DenunciationInterest};
+use massa_models::slot::Slot;
+use massa_models::timeslots::get_closest_slot_to_timestamp;
+use massa_time::MassaTime;
+
+/// Structure gathering all elements needed by the factory thread
+pub(crate) struct DenunciationFactoryWorker {
+    cfg: FactoryConfig,
+    channels: FactoryChannels,
+    /// Factory manager command receiver
+    factory_receiver: Receiver<()>,
+    consensus_receiver: Receiver<DenunciationInterest>,
+    endorsement_pool_receiver: Receiver<DenunciationInterest>,
+    /// Internal cache for endorsement denunciation
+    /// store at most 1 endorsement per entry, as soon as we have 2 we produce a Denunciation
+    endorsements_by_slot_index: HashMap<(Slot, u32), EndorsementDenunciationStatus>,
+    /// Internal cache for block header denunciation
+    block_header_by_slot: HashMap<Slot, BlockHeaderDenunciationStatus>,
+}
+
+impl DenunciationFactoryWorker {
+    /// Creates the `FactoryThread` structure to gather all data and references
+    /// needed by the factory worker thread.
+    pub(crate) fn spawn(
+        cfg: FactoryConfig,
+        channels: FactoryChannels,
+        factory_receiver: Receiver<()>,
+        consensus_receiver: Receiver<DenunciationInterest>,
+        endorsement_pool_receiver: Receiver<DenunciationInterest>,
+    ) -> thread::JoinHandle<()> {
+        thread::Builder::new()
+            .name("denunciation-factory".into())
+            .spawn(|| {
+                let mut factory = Self {
+                    cfg,
+                    channels,
+                    factory_receiver,
+                    consensus_receiver,
+                    endorsement_pool_receiver,
+                    endorsements_by_slot_index: Default::default(),
+                    block_header_by_slot: Default::default(),
+                };
+                factory.run();
+            })
+            .expect("failed to spawn thread : denunciation-factory")
+    }
+
+    /// Process new secured header (~ block header)
+    fn process_new_secured_header(
+        &mut self,
+        block_header_denunciation_interest: DenunciationInterest,
```

Comment:
I'd rather keep the enum as it is because:
* It is used directly in internal caching
* There is a Denunciation try_from impl for DenunciationInterest and not for each particular variant

Reply:
ok then let's keep it, it's not very important